source/plugins/go/src/ingestion_token_utils.go

```diff
@@ -0,0 +1,430 @@
+	// Create a new request using http
+	req, err := http.NewRequest("GET", amcs_endpoint.String(), nil)
+	if err != nil {
+		message := fmt.Sprintf("Error creating HTTP request for AMCS endpoint: %s", err.Error())
+		Log(message)
+		return configurationId, channelId, err
+	}
+	req.Header.Set("Authorization", bearer)
+
+	var resp *http.Response = nil
+
+	for i := 0; i < 3; i++ {	
+		resp, err = HTTPClient.Do(req)
+		if err != nil {
+			message := fmt.Sprintf("Error calling AMCS endpoint: %s", err.Error())
+			Log(message)
+			SendException(message)
+			continue
+		}
+		if resp != nil && resp.Body != nil {
+			defer resp.Body.Close() 
+	    }
+		Log("getAgentConfiguration Response Status: %d", resp.StatusCode)
+		if resp.StatusCode != 200 {
+			message := fmt.Sprintf("getAgentConfiguration Request failed with an error code : %d", resp.StatusCode)
+			Log(message)
+			SendException(message)
+			continue
+		} 
+		break // call succeeded, don't retry any more
+	}
+	if resp == nil {
+		Log("getAgentConfiguration Request ran out of retries")
+		return configurationId, channelId, err
+	}
+	responseBytes, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		Log("Error reading response body from AMCS API call : %s", err.Error())
+		return configurationId, channelId, err
+	}
+
+	// Unmarshall response body into struct
+	var agentConfiguration AgentConfiguration
+	err = json.Unmarshal(responseBytes, &agentConfiguration)
+	if err != nil {
+		Log("Error unmarshalling the response: %s", err.Error())
+		return configurationId, channelId, err
+	}
+
+	if len(agentConfiguration.Configurations) == 0 {
+		Log("Received empty agentConfiguration.Configurations array")
+		return configurationId, channelId, err
+	}
+
+	if len(agentConfiguration.Configurations[0].Content.Channels) == 0 {
+		Log("Received empty agentConfiguration.Configurations[0].Content.Channels")
+		return configurationId, channelId, err
+	}
+
+	configurationId = agentConfiguration.Configurations[0].Configurationid
+	channelId = agentConfiguration.Configurations[0].Content.Channels[0].ID     
+
+	Log("obtained configurationId: %s, channelId: %s", configurationId, channelId)
+	Log("Info getAgentConfiguration: end")
+
+	return configurationId, channelId, nil
+}
+
+func getIngestionAuthToken(imdsAccessToken string, configurationId string, channelId string) (ingestionAuthToken string, refreshInterval int64, err error) {
+	Log("Info getIngestionAuthToken: start")
+	ingestionAuthToken = ""
+	refreshInterval = 0
+	var amcs_endpoint *url.URL
+	osType := os.Getenv("OS_TYPE")
+	resourceId := os.Getenv("AKS_RESOURCE_ID")
+	resourceRegion := os.Getenv("AKS_REGION")
+	MCS_ENDPOINT := os.Getenv("MCS_ENDPOINT")
+	apiVersion := "2020-04-01-preview"
+	amcs_endpoint_string := fmt.Sprintf("https://%s.handler.control."+MCS_ENDPOINT+"%s/agentConfigurations/%s/channels/%s/issueIngestionToken?platform=%s&api-version=%s", resourceRegion, resourceId, configurationId, channelId, osType, apiVersion)
```

Comment:
[very pedantic] `amcs_endpoint_string := fmt.Sprintf("https://%s.handler.control.%s%s/agentConfigurations/%s/channels/%s/issueIngestionToken?platform=%s&api-version=%s", resourceRegion, MCS_ENDPOINT, resourceId, configurationId, channelId, osType, apiVersion)` would be a lot cleaner

Reply:
fixed.